### PR TITLE
[inductor] force 'fork' method for processes, cleanup

### PR DESCRIPTION
**Pull request by zdevito**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #87411

To cooperate with other multithreading methods, this
forces the process pool to use 'fork' even if others have set it
diferently. We require fork because otherwise `if __name__ == __main__`
needs to be set which we do not control as a library.

Furthermore this adds code to cleanup worker processes if
the parent exits abnormally (e.g. segfault). Previously we would leave
live but inactive workers around.

cc @jansel @lezcano @fdrocha @mlazos @soumith @voznesenskym @yanboliang

### CODE REVIEW

File: torch/_inductor/codecache.py [hunk 1]
```diff
@@ -279,7 +282,31 @@ def process_pool():
...
+        # if this process dies abnormally (e.g. segfault)
+        # it will not shut down the workers. Instead
+        # the workers will have their parent reassigned to the
+        # init process. This launches a separate thread to
+        # watch for the worker getting reassigned,
+        # and cleans it up in this case.
+        def init():
+            def run():
+                while True:
+                    if orig_ppid != os.getppid():
```
Comment (soumith) [inline torch/_inductor/codecache.py:296]: a big try/catch around this to just ignore any exceptions that arise (probably during the main process's shutdown)?

**Review by soumith — APPROVED** [review]

File: torch/_inductor/codecache.py (hunk 1 above)
Comment (zdevito) [reply torch/_inductor/codecache.py:296]: What sequence of events could arise that catching an exception will resolve? os.getpid and os.getppid can't throw exceptions,  the underlying unix calls do not error. os.kill can, but if we reach that code, we want the current process to die, and if the kill command doesn't work, we don't have a way to kill the process anyway.

**Review by soumith — APPROVED** [review]

File: torch/_inductor/codecache.py (hunk 1 above)
Comment (soumith) [reply torch/_inductor/codecache.py]: idk, i guess the comment was just to be preemptive. Didn't really think too much about it.

**Review by anijain2305 — APPROVED** [review]

**Comment by pritamdamania87** [comment]
@zdevito @soumith @anijain2305 Wondering why we used "fork" here instead of "spawn". "fork" is not safe in a multithreaded environment, see: https://www.microsoft.com/en-us/research/uploads/prod/2019/04/fork-hotos19.pdf and https://discuss.python.org/t/switching-default-multiprocessing-context-to-spawn-on-posix-as-well/21868.

In a multithreaded environment "fork" can cause deadlocks and segfaults. We are seeing a deadlock due to the use of "fork" here and need to disable inductor workers as a result.

**Comment by zdevito** [comment]
We can't use "spawn" because it expects that the top-level launcher script for any program a user wrote is has a `if __name__ == "__main__"` line. Otherwise it will re-execute the user script in all the spawned processes. Better solution here would be to Popen a separate script entirely to run the compilation threads, but then we have rebuild some of the infra that the multiprocessing framework provides communicating with workers.

**Comment by pritamdamania87** [comment]
> We can't use "spawn" because it expects that the top-level launcher script for any program a user wrote is has a if __name__ == "__main__" line.

Isn't it ok to expect the top-level launcher to specify this since it is a well known idiom? If not specified, I believe you do see an appropriate error message like this:

```
RuntimeError: 
        An attempt has been made to start a new process before the
        current process has finished its bootstrapping phase.

        This probably means that you are not using fork to start your
        child processes and you have forgotten to use the proper idiom
        in the main module:

            if __name__ == '__main__':
                freeze_support()
                ...

        The "freeze_support()" line can be omitted if the program
        is not going to be frozen to produce an executable.
```

On the other hand with "fork" this causes non-deterministic segfaults and deadlocks which are extremely hard to debug and root cause.

**Comment by eellison** [comment]
@pritamdamania87 I'm looking into alternatives here, hope to have fix in soon. The alternatives should work for your use case.

**Comment by pritamdamania87** [comment]
Thanks a lot @zdevito @eellison, really appreciate it!

**Pull request closed without merging**